server/pulp/plugins/file/distributor.py

```diff
@@ -212,9 +213,13 @@ def _symlink_unit(self, build_dir, unit, target_paths):
         :param target_paths: The list of paths the unit should be symlinked to.
         :type  target_paths: list of L{str}
         """
+        cwd = os.path.abspath(os.path.curdir)
         for target_path in target_paths:
-            # symlink_filename = os.path.join(build_dir, unit.unit_key['name'])
-            symlink_filename = os.path.join(build_dir, target_path)
+            # any two absolute paths always share at least the root element
+            abs_target_path = os.path.abspath(target_path)
+            target_path_cwd_prefix = os.path.commonprefix([cwd, os.path.dirname(abs_target_path)])
+            confined_target_path = abs_target_path.replace(target_path_cwd_prefix, '').lstrip('/')
```

Comment:
This is buggy. Consider what happens to `/var../one/up` (if `cwd`start with "/var") or `/../b/o/g/u/s`.

If  we really want to support relative paths, I am thinking of the following checks:
- target_path must not be absolute
- target_path must not end with `/` (indicating a directory, for which we can't have a  valid digest or length)
- os.path.normpath(target_path) must not begin with `../` (which indicates that the relative path leaves the working directory)
- os.path.normpath(target_path) must not be `..` (which indicates that the relative path leaves the working directory)

A path that violates these requirement should raise an error. 

If such a path happens to work with the current implementation, it writes a file at a location outside of the build directory and the resulting published repo is broken as this file is not within the published tree (and has a manifest file pointing to a non-existent or wrong location).

I think it is justifiable to raise an error in this case.

Reply:
@gmbnomis the bug is in my `replace` use; it has to be limited to a single occurence. Your example paths then get confined as expected:
```
In [1]: import os                                                                                                                                                                                          [2/1679]
                                                                                                                                                                                                                   
In [2]: cwd = os.path.abspath(os.path.curdir)                                                                                                                                                                      

In [3]: cwd                                                                                        
Out[3]: '/home/vagrant/devel/pulp'                        
                                                        
In [4]: cd /var/                                                                
/var                                        
            
In [5]: cwd = os.path.abspath(os.path.curdir) 
                                         
In [6]: cwd                                                                     
Out[6]: '/var'                                                      
                                                                            
In [7]: target_path = '/var/../one/up'                     
                                                                                                   
In [8]: abs_target_path = os.path.abspath(target_path)                                             
                                                                             
In [9]: abs_target_path                                                              
Out[9]: '/one/up'                                                                                  
                                                                                                    
In [10]: target_path_cwd_prefix = os.path.commonprefix([cwd, os.path.dirname(abs_target_path)])
                                     
In [11]: target_path_cwd_prefix
Out[11]: '/'

In [12]: confined_target_path = abs_target_path.replace(target_path_cwd_prefix, '').lstrip('/')
                                    
In [13]: confined_target_path                                                                                                                                                                                     
Out[13]: 'oneup'                                                               
                                              
In [14]: abs_target_path.replace(target_path_cwd_prefix, '')  
Out[14]: 'oneup'
                                                
In [15]: abs_target_path.replace(target_path_cwd_prefix, '', 1)
Out[15]: 'one/up'            
                                      
In [16]: abs_target_path.replace(target_path_cwd_prefix, '', 1).lstrip('/')
Out[16]: 'one/up'
                 
In [17]: target_path = '/../b/o/g/u/s'
                                             
In [18]: abs_target_path = os.path.abspath(target_path)
           
In [19]: target_path_cwd_prefix = os.path.commonprefix([cwd, os.path.dirname(abs_target_path)])

In [20]: confined_target_path = abs_target_path.replace(target_path_cwd_prefix, '', 1).lstrip('/')
    
In [21]: confined_target_path
Out[21]: 'b/o/g/u/s'
```
>  A path that violates these requirement should raise an error.

I'd prefer this sort of checking happened during the import of metadata rather than during the publishing of the repository.

> if such a path happens to work with the current implementation, it writes a file at a location outside of the build directory and the resulting published repo is broken as this file is not within the published tree (and has a manifest file pointing to a non-existent or wrong location).
>
> I think it is justifiable to raise an error in this case.

I thought it would be simpler to assume any absolute `target_path` meaning a `chroot` into the repository i.e treat both `/foo/bar/baz` and `foo/bar/baz` the same

Reply:
> In [7]: target_path = '/var/../one/up'

My example was:

`target_path = '/var../one/up'`

> > A path that violates these requirement should raise an error.
> 
> I'd prefer this sort of checking happened during the import of metadata rather than during the publishing of the repository.

I agree on the import part. From a security point of view, it is best to reject wrong data when entering the system. But my conclusion fpr publication is different: If we have proper rejection at importing/syncing, there is no need for publication to silently drop errors by forcing invalid paths into the build_dir. To the contrary, this should result in an error as something with the validation is very wrong in this case.

> 
> > if such a path happens to work with the current implementation, it writes a file at a location outside of the build directory and the resulting published repo is broken as this file is not within the published tree (and has a manifest file pointing to a non-existent or wrong location).
> > I think it is justifiable to raise an error in this case.
> 
> I thought it would be simpler to assume any absolute `target_path` meaning a `chroot` into the repository i.e treat both `/foo/bar/baz` and `foo/bar/baz` the same

But if you want that, why don't you do exactly that: Treat  `/foo/bar/baz` as `foo/bar/baz` (i.e. get rid of leading slashes) and treat it as a relative path into `build_dir`. As said, `target_path` has nothing to do with the current working directory IMHO.

**But:** This is a new feature changing the semantics of the "name" attribute (and we must look into the "name"s written into PULP_MANIFEST as well.) Additionally, the "../some_file" -> "<build_dir>/some_file" case still makes no sense to me.

Currently, almost all absolute paths lead to an error when publishing (because the location is usually not writable). With your change, they have a well defined meaning. I don't think this is the right place to introduce such a feature. Especially, if the next change you plan it to reject absolute paths at import/sync.

Reply:
>> I'd prefer this sort of checking happened during the import of metadata rather than during the publishing of the repository.

> I agree on the import part. From a security point of view, it is best to reject wrong data when entering the system. But my conclusion for publication is different: If we have proper rejection at importing/syncing, there is no need for publication to silently drop errors by forcing invalid paths into the build_dir. To the contrary, this should result in an error as something with the validation is very wrong in this case.

With this patch I'm not silently dropping errors of invalid paths, I'm rather introducing a convention.

> My example was:
>
> target_path = '/var../one/up'

Sorry I assumed it was a typo; I guess the resulting path is OK nevertheless: `/var/lib/pulp/build_dir/my_repo/var../one/up`

> But: This is a new feature changing the semantics of the "name" attribute (and we must look into the "name"s written into PULP_MANIFEST as well.)

I agree, it worries me a bit; if we have users with correct absolute path in the manifest, this patch will break them. So will the publish-time validation.

> Additionally, the "../some_file" -> "<build_dir>/some_file" case still makes no sense to me.

wit this patch it's the same semantics as `$ cd / ; ls ../mnt` it's odd but not broken 

> Currently, almost all absolute paths lead to an error when publishing (because the location is usually not writable). With your change, they have a well defined meaning. I don't think this is the right place to introduce such a feature. Especially, if the next change you plan it to reject absolute paths at import/sync.

Even then, I'm not preventing publishing of what was allowed to be imported. I find this a slightly better option.